File: 0007-reverse-integer/0007-reverse-integer.py
class Solution:
    def reverse(self, x):
        INT_MIN = -2**31       # -2147483648
        INT_MAX = 2**31 - 1    # 2147483647

        result = 0
        sign = -1 if x < 0 else 1
        x = abs(x)

        while x != 0:
            digit = x % 10
            x //= 10

            # Check overflow
            if result > (INT_MAX - digit) // 10:
                return 0

            result = result * 10 + digit

        return sign * result

File: 0007-reverse-integer/0007-reverse-integer.cpp
class Solution {
public:
    int reverse(int x) {
        long reversed = 0;  // use long to detect overflow

        while (x != 0) {
            int digit = x % 10;   // get last digit
            x /= 10;              // remove last digit

            reversed = reversed * 10 + digit;

            // check for overflow before returning
            if (reversed > INT_MAX || reversed < INT_MIN)
                return 0;
        }

        return static_cast<int>(reversed);
    }
};
